### PR TITLE
MOO-71 Adding patient trend results widget

### PR DESCRIPTION
**Pull request by Imran536**
MOO-71 Adding patient trend results widget
Though depends on pull request in MSF-core https://github.com/jembi/openmrs-module-msfcore/pull/114

### CODE REVIEW

File: omod/src/main/web/dashboardwidgets/index.js [hunk 1]
```diff
@@ -9,9 +8,10 @@ import ProgramStatistics from './programstatistics';
 import ProgramStatus from './programstatus';
 import Relationships from './relationships';
 import VisitByEncounterType from './visitbyencountertype';
+import ResultsTrendView from './resultstrendview'
 
 export default angular.module("openmrs-contrib-dashboardwidgets", [ DataIntegrityViolations, LatestObsForConceptList,
-    ObsAcrossEncounters, ObsGraph, Programs, ProgramStatistics, ProgramStatus, Relationships, VisitByEncounterType]).name;
+    ObsAcrossEncounters, ObsGraph, Programs, ProgramStatistics, ProgramStatus, Relationships, VisitByEncounterType,ResultsTrendView]).name;
```
Comment (lauravignoli) [inline omod/src/main/web/dashboardwidgets/index.js]: @Imran536 There's a comma missing here, please add it.

File: omod/src/main/web/dashboardwidgets/resultstrendview/index.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import angular from 'angular';
+import openmrsApi from '@openmrs/angularjs-openmrs-api';
+import commons from './../dashboardwidgets.services';
+import { ResultsTrendViewComponent } from './resultstrendview.component';
+
+export default angular.module("openmrs-contrib-dashboardwidgets.testresults71", [openmrsApi, commons])
```
Comment (lauravignoli) [inline omod/src/main/web/dashboardwidgets/resultstrendview/index.js]: @Imran536 What is this name related to? Maybe something that you left by mistake? testresults71

File: omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		 		<option value=7>Last week</option>
+		 		<option value=14>Last two weeks</option>
+				<option value=90>Last three months</option>
+ 			</select>
+ 			<br><br>
+ 			<a class="service" href="#" ng-click="$ctrl.generateVitalsChart()">Generate chart</a>	     	 	
+     	</td>
+		
+	</tr>
+<!-- 	<tr>
```
Comment (lauravignoli) [inline omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html]: @Imran536 Any reason why this was left commented out?

File: omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    </ui-select-choices>
+  </ui-select>
+  
+<div style="padding-bottom: 10px">
+        <strong>
+        	Vitals Chart
+        </strong>
+    </div>
+<div ng-if="$ctrl.vitalChartLabels.length > 0">
+    <!-- <div style="padding-bottom: 10px">
```
Comment (lauravignoli) [inline omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html]: @Imran536 Any reason why this was left commented out?

File: omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+<div style="padding-bottom: 10px">
+        <strong>
+        	Lab Results Chart
+        </strong>
+    </div>
+
+	<br/>
+
+<div ng-if="$ctrl.labChartLabels.length > 0">
+    <!-- <div style="padding-bottom: 10px">
```
Comment (lauravignoli) [inline omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html]: @Imran536 Any reason why this was left commented out?

File: omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.controller.js [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+                            console.log(i);
+                        }
+                    }
+                }
+            }
+    	}
+    }
+    
+    fetchVitalsConcepts() {
+        this.openmrsRest.getFull("concept/" + '106411d9-f354-4160-8603-4dec659d1bb6', {v: 'custom:(uuid,display,names:(display,conceptNameType)'}).then((concept) => {
```
Comment (lauravignoli) [inline omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.controller.js]: @Imran536 The concept sets using in this widget should be configurable.

File: omod/src/main/web/dashboardwidgets/index.js (hunk 1 above)
Comment (Imran536) [reply omod/src/main/web/dashboardwidgets/index.js]: @lauravignoli I suppose you are talking about the missing semi-colon after the last import statement. I have added it.

File: omod/src/main/web/dashboardwidgets/resultstrendview/index.js (hunk 2 above)
Comment (Imran536) [reply omod/src/main/web/dashboardwidgets/resultstrendview/index.js]: @lauravignoli yes, it was left by mistake. I have changed the name to **resultstrendview**.

File: omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.controller.js (hunk 6 above)
Comment (Imran536) [reply omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.controller.js]: @lauravignoli This has already been changed in the latest commit. Here is the link:
https://github.com/InteractiveHealthSolutions/openmrs-module-coreapps/blob/71-trend-results-widget/omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.controller.js

File: omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html (hunk 4 above)
Comment (Imran536) [reply omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html]: @lauravignoli This has already been changed in the latest commit. Here is the link:
https://github.com/InteractiveHealthSolutions/openmrs-module-coreapps/blob/71-trend-results-widget/

File: omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html (hunk 5 above)
Comment (Imran536) [reply omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html]: @lauravignoli This has already been changed in the latest commit. 
https://github.com/InteractiveHealthSolutions/openmrs-module-coreapps/tree/71-trend-results-widget

File: omod/src/main/resources/apps/dashboardWidgets_app.json [hunk 7]
```diff
@@ -23,6 +22,35 @@
...
+   "instanceOf": "coreapps.template.dashboardWidget",
+   "description": "coreapps.resultstrendview.app.description",
+   "order": 10,
+   "config": {
+     "widget": "resultstrendview",
+     "icon": "",
+     "label": "RESULTS TREND VIEW",
+     "maxResults": "100",
+     "maxAge": "1y",
+     "vitalsListUuid":"106411d9-f354-4160-8603-4dec659d1bb6",
```
Comment (lauravignoli) [inline omod/src/main/resources/apps/dashboardWidgets_app.json]: @Imran536 I think I forgot to add a comment here. These uuids are specific to the msf project, there's no set for vitals by default. I think this should be left blank and explain how to be used by documentation.

What do you think the best approach should be @ssmusoke

File: omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html (hunk 3 above)
Comment (Imran536) [reply omod/src/main/web/dashboardwidgets/resultstrendview/resultstrendview.html]: @lauravignoli This has already been changed in the latest commit.
https://github.com/InteractiveHealthSolutions/openmrs-module-coreapps/tree/71-trend-results-widget

File: omod/src/main/resources/apps/dashboardWidgets_app.json (hunk 7 above)
Comment (ssmusoke) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: I think the default widget should use the vitals uuids available in Ref App, then in msfcore it can be configured to use msf specific uuids

File: omod/src/main/resources/apps/dashboardWidgets_app.json [hunk 8]
```diff
@@ -23,6 +22,35 @@
...
+     "labResultsUuid":"23281464-74d8-47d9-9a39-7a1f1d7caa4f",
+     "links":{
+     "Go to Page":"/openmrs/coreapps/patientdashboard/resultsTrendView.page?patientId={{patientUuid}}&app=coreapps.resultstrendview"
```
Comment (ssmusoke) [inline omod/src/main/resources/apps/dashboardWidgets_app.json]: I think returnUrl is the parameter used commonly instead of "Go to Page"

File: omod/src/main/resources/apps/dashboardWidgets_app.json (hunk 7 above)
Comment (lauravignoli) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: @ssmusoke This widget uses a set for vitals and a set for lab results. The only thing that is part of ref app is the vitals concepts, but not the vitals set, or the lab results set or the lab results concepts.

Do you think it would be a good approach to add this information as documentation but have this values empty by default?

Comment (ssmusoke) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: @lauravignoli Just a thought what is the difference in processing of the two? Why not conceptSetUuid also allowing a comma delimited list of uuids

Comment (lauravignoli) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: So that if more lab results are added in future they display here by default? But if this is not a common practice, we can change it.

Comment (ssmusoke) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: Exactly:

1. The common practice is to provide a comma delimited list of concepts which I think is the way to go as that list can be extended. 

2. If you combine vitals and the lab results in a single value then it will be much easier to use and configure

Comment (lauravignoli) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: 1. The common practice is to provide a comma delimited list of concepts which I think is the way to go as that list can be extended" **=> but how can we do this by config? With the set, we can just add set members right? If the application goes live, there's no way to update the list of concepts unless updating the code?**

2. If you combine vitals and the lab results in a single value then it will be much easier to use and configure => if vitals and lab results were together we would need to separate them to display the 2 dropdowns, how would we do that?

File: omod/src/main/web/dashboardwidgets/index.js (hunk 1 above)
Comment (lauravignoli) [reply omod/src/main/web/dashboardwidgets/index.js]: @Imran536 Sorry I meant a missing space!!

File: omod/src/main/resources/apps/dashboardWidgets_app.json (hunk 7 above)
Comment (lauravignoli) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: Ah ok, I didn't think of that, we can disable the existing widget and create a new app with the new concepts. 

@Imran536 I think after this conversation, the best approach is to use a comma delimited list of concept sets and/or concept Ids for each list. What do you think?

Comment (Imran536) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: @lauravignoli We will discuss it and get back to you. For now, we are focusing on full patient summary story.

File: omod/src/main/resources/apps/dashboardWidgets_app.json (hunk 8 above)
Comment (Imran536) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: @ssmusoke This "Go to Page" is the text of the link.

File: omod/src/main/resources/apps/dashboardWidgets_app.json [hunk 9]
```diff
@@ -120,5 +148,33 @@
         }
       }
     ]
+  },
```
Comment (lauravignoli) [inline omod/src/main/resources/apps/dashboardWidgets_app.json]: @Imran536 Why has this been updated?

File: omod/src/main/resources/apps/dashboardWidgets_app.json (hunk 7 above)
Comment (lauravignoli) [reply omod/src/main/resources/apps/dashboardWidgets_app.json]: @Imran536 any updates on this?

File: omod/src/main/webapp/pages/patientdashboard/resultsTrendView.gsp [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    var breadcrumbs = [
+        { icon: "icon-home", link: '/' + OPENMRS_CONTEXT_PATH + '/index.htm' },
+        { label: "${ ui.escapeJs(ui.encodeHtmlContent(ui.format(patient.patient))) }" ,
+            link: '${ ui.urlBind("/" + contextPath + dashboardUrl, [ patientId: patient.patient.id ] ) }'}
+    ];
+
+    jq(function(){
+        jq(".tabs").tabs();
+
+        // make sure we reload the page if the location is changes; this custom event is emitted by by the location selector in the header
```
Comment (tmvumbi2) [inline omod/src/main/webapp/pages/patientdashboard/resultsTrendView.gsp]: @Imran536 small typos:  ... the location is **changed**;  ... is emitted by the ... (remove one 'by')

File: omod/src/main/webapp/pages/patientdashboard/resultsTrendView.gsp [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        jq(document).on('sessionLocationChanged', function() {
+            window.location.reload();
+        });
+    });
+
+    var patient = { id: ${ patient.id } };
```
Comment (tmvumbi2) [inline omod/src/main/webapp/pages/patientdashboard/resultsTrendView.gsp]: @Imran536 the controller injected `userId` to the page model (can be used to replace `patient.id`)

File: omod/src/main/java/org/openmrs/module/coreapps/page/controller/patientdashboard/ResultsTrendViewPageController.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+                    @RequestParam(value = "visitId", required = false) Visit visit,
+                    @InjectBeans PatientDomainWrapper patientDomainWrapper,
+                    @SpringBean("orderService") OrderService orderService,
+                    @SpringBean("adtService") AdtService adtService,
+                    @SpringBean("appFrameworkService") AppFrameworkService appFrameworkService,
+                    @SpringBean("coreAppsProperties") CoreAppsProperties coreAppsProperties,
+                    @SpringBean("applicationEventService") ApplicationEventService applicationEventService,
+                    UiSessionContext sessionContext) {
+    	
+        model.addAttribute("patient", patient);
```
Comment (tmvumbi2) [inline omod/src/main/java/org/openmrs/module/coreapps/page/controller/patientdashboard/ResultsTrendViewPageController.java]: @Imran536 The attribute 'patient' is set (updated?) on line 54, do we need to set it here as well?
